doc/tincd.8
.TH TINCD 8 "June 2000" "tinc version 1.0pre3" "FSF"
.SH NAME
tinc \- tinc VPN daemon
.SH SYNOPSIS
.B tincd
[\fIoption\fR]...
.SH DESCRIPTION
.PP

This is the daemon of tinc, a secure virtual private
network (VPN) project. When started, tincd will read
it's configuration file to determine what virtual subnets
it has to serve and to what other tinc daemons it should connect.
It will connect to the ethertap or tun/tap device and set up a socket
for incomming connections.
If that succeeds, it will detach from the controlling terminal and
continue in the background, accepting and setting up connections to other
tinc daemons that are part of the virtual private network.

.SH OPTIONS
.TP
\fB\-c\fR, \fB\-\-config\fR=\fIDIR\fR
Read configuration options from DIR.
.TP
\fB\-D\fR, \fB\-\-no\-detach\fR
Don't fork and detach. This will also disable the automatic
restart mechanism for fatal errors.
.TP
\fB\-d\fR
Increase debug level (see below).
.TP
\fB\-k\fR, \fB\-\-kill\fR
Attempt to kill a running tincd and exit.
.TP
\fB\-n\fR, \fB\-\-net\fR=\fINETNAME\fR
Connect to net NETNAME.
.TP
\fB\-K\fR, \fB\-\-generate-key\fR[=\fIBITS]\fR
Generate public/private RSA keypair and exit. If BITS is omitted,
the default length will be 1024 bits.
.TP
\fB\-\-help\fR
Display short list of options.
.TP
\fB\-\-version\fR
Output version information and exit.
.PP
.SH "SIGNALS"
.TP
\fBHUP\fR
Closes all connections, rereads the configuration file and restarts the daemon.
.TP
\fBINT\fR
Closes all connections and quits.
.TP
\fBUSR1\fR
Dumps the connection list to syslog.
.TP
\fBUSR2\fR
Dumps the subnet list to syslog.
.TP
\fBALRM\fR
Forces tincd to try to connect to an uplink immediately. Usually tincd attempts
to do this itself, but increases the time it waits between the attempts each time
it failed, and if tincd didn't succeed to connect to an uplink the first time after
it started, it defaults to the maximum time of 15 minutes.
.PP
.SH "DEBUG LEVELS"
The tinc daemon can send a lot of messages to the syslog. The more \fB\-d\fR options are
given to tincd, the more messages it will log. Each level inherits all messages of the
previous level:
.TP
\fIno debug options\fR
This will log a message indicating tincd has started along with a version number.
It will also any serious error.
.TP
\fB\-d\fR
This will log all connections that are made with other tinc daemons.
.TP
\fB\-dd\fR
This will log status and error messages from other tinc daemons.
.TP
\fB\-ddd\fR
This will log all requests that are exchanged with other tinc daemons. These include
authentication, key exchange and connection list updates.
.TP
\fB\-dddd\fR
This will log a copy of everything received on the meta socket.
.TP
\fB\-ddddd\fR
This will log all network traffic over the virtual private network.
.PP
.SH "FILES"
.TP
\fI/etc/tinc/<NETNAME>/tinc.conf\fR
The configuration file for tincd. This should also contain the private RSA key.
.TP
\fI/etc/tinc/<NETNAME>/hosts/*\fR
The directory containing the host configuration files
used to authenticate other tinc daemons. They contain
the public RSA keys of other hosts.
.PP
.SH "BUGS"
Maintaining a connection list on each tinc daemon that can connect and disconnect at any
moment, and making sure that all connections satisfy the tree property isn't easy. Although
we have done a lot to make sure tinc is sturdy and foolproof, it might happen that
some connection lists get corrupted.
.PP
\fBThe cryptography in tinc is not well tested yet. Use it at your own risk!\fR
.PP
If you find any bugs, report them to tinc@nl.linux.org.
.PP
.SH "TODO"
A lot, especially security auditting.
.PP 
.SH "SEE ALSO"
\fBtinc.conf\fR(5)
.TP
\fBhttp://tinc.nl.linux.org/\fR
.TP
\fBhttp://www.cabal.org/\fR
.PP
The full documentation for
.B tinc
is maintained as a Texinfo manual.  If the
.B info
and
.B tinc
programs are properly installed at your site, the command
.IP
.B info tinc
.PP
should give you access to the complete manual.
.PP
tinc comes with ABSOLUTELY NO WARRANTY.  This is free software,
and you are welcome to redistribute it under certain conditions;
see the file COPYING for details.
.SH "AUTHORS"
.na
.nf
Ivo Timmermans <itimmermans@bigfoot.com>
Guus Sliepen <guus@sliepen.warande.net>

And thanks to many others for their contributions to tinc!
.PP


NEWS
Version 1.0.37               March 1 2026

 * Minor fixes.
 * Drop support for Linux ethertap devices.
 * Ensure tinc compiles cleanly on modern systems.

This will most likely be the last version of tinc 1.0.

Version 1.0.36               August 26 2019

 * Fix compiling tinc with certain versions of the OpenSSL library.
 * Fix parsing some IPv6 addresses with :: in them.
 * Fix GraphDumpFile output to handle node names starting with a digit.
 * Fix a potential segmentation fault when fragmenting packets.

Thanks to Rosen Penev, Quentin Rameau and Werner Schreiber for their
contributions to this version of tinc.

Version 1.0.35               October 5 2018

 * Prevent oracle attacks (CVE-2018-16737, CVE-2018-16738).
 * Prevent a MITM from forcing a NULL cipher for UDP (CVE-2018-16758).
 * Minor fixes in the documentation.

Thanks to Amine Amri and Rafael Sadowski for their contributions to this
version of tinc.

Version 1.0.34               June 12 2018

 * Fix a potential segmentation fault when connecting to an IPv6 peer via a
   proxy.
 * Minor improvements to the build system.
 * Make the systemd service file identical to the one from the 1.1 branch.
 * Fix a potential problem causing IPv4 sockets to not work on macOS.

Thanks to Maximilian Stein and Wang Liu Shuai for their contributions to this
version of tinc.

Version 1.0.33               November 4 2017

 * Allow compilation from a build directory.
 * Source code cleanups.
 * Fix some options specified on the command line not surviving a HUP signal.
 * Handle tun/tap device returning EPERM or EBUSY.
 * Disable PMTUDiscovery when TCPOnly is used.
 * Support the --runstatedir option of the autoconf 2.70.

Thanks to Rafael Sadowski and Pierre-Olivier Mercier for their contributions to
this version of tinc.

Version 1.0.32               September 2 2017

 * Fix segmentation fault when using Cipher = none.
 * Fix Proxy = exec.
 * Support PriorityInheritance for IPv6 packets.
 * Fixes for Solaris tun/tap support.
 * Bind outgoing TCP sockets when ListenAddress is used.

Thanks to Vittorio Gambaletta for his contribution to this version of tinc.

Version 1.0.31               January 15 2017

 * Remove ExecStop in tinc@.service.

Thanks to Élie Bouttier for his contribution to this version of tinc.

Version 1.0.30               October 30 2016

 * Fix troubles connecting to some HTTP proxies.

 * Add mitigations for the Sweet32 attack when using a 64-bit block cipher.

 * Use AES256 and SHA256 as the default encryption and digest algorithms.

Version 1.0.29               October 9 2016

 * Fix UDP communication with peers with link-local IPv6 addresses.

 * Ensure compatibility with OpenSSL 1.1.0.

 * Ensure autoreconf can be run without requiring autoconf-archive.

 * Log warnings about dropped packets only at debug level 5.

Version 1.0.28               April 10 2016

 * Fix compilation on BSD platforms.

 * Add systemd service files.

Version 1.0.27               April 10 2016

 * When using Proxy, let the proxy resolve hostnames if tinc can't.

 * Fixes and improvements of the DecrementTTL option.

 * Fixed the $NAME variable in subnet-up/down scripts for the local Subnets.

 * Fixed potentially wrong checksum generation when clamping the MSS.

 * Properly choose between the system's or our own copy of getopt.

 * Fixed compiling tinc for Cygwin with MinGW installed.

 * Added support for OS X utun interfaces.

 * Documentation updates and minor fixes.

Thanks to Vittorio Gambaletta, LunarShaddow, Florian Weik and Nathan Stratton
Treadway for their contributions to this version of tinc.

Version 1.0.26               July 5 2015

 * Tinc now forces glibc to reload /etc/resolv.conf for every hostname lookup.

 * Fixed --logfile without a filename on Windows.

 * Ensure tinc can be compiled when using musl libc.

Thanks to Jo-Philipp Wich for his contribution to this version of tinc.

Version 1.0.25               December 22 2014

 * Documentation updates.

 * Support linking against -lresolv on Mac OS X.

 * Fix scripts on Windows when using the ScriptsInterpreter option.

 * Allow a minimum reconnect timeout to be specified.

 * Support PriorityInheritance on IPv6 sockets.

Thanks to David Pflug, Baptiste Jonglez, Alexis Hildebrandt, Borg, Jochen Voss,
Tomislav Čohar and VittGam for their contributions to this version of tinc.

Version 1.0.24               May 11 2014

 * Various compiler hardening flags are enabled by default.

 * Updated support for Solaris, allowing switch mode on Solaris 11.

 * Configuration will now also be read from a conf.d directory.

 * Various updates to the documentation.

 * Tinc now forces glibc to reload /etc/resolv.conf after it receives SIGALRM.

 * Fixed a potential routing loop when IndirectData or TCPOnly is used and
   broadcast packets are being sent.

 * Improved security with constant time memcmp and stricter use of OpenSSL's
   RNG functions.

 * Fixed all issues found by Coverity.

Thanks to Florent Clairambault, Vilbrekin, luckyhacky, Armin Fisslthaler, Loïc
Dachary and Steffan Karger for their contributions to this version of tinc.

Version 1.0.23               October 19 2013

 * Start authentication immediately on outgoing connections (useful for sslh).

 * Fixed segfault when Name = $HOST but $HOST is not set.

 * Updated the build system and the documentation.

 * Clean up child processes left over from Proxy = exec.

Version 1.0.22               August 13 2013

 * Fixed the combination of Mode = router and DeviceType = tap.

 * The $NAME variable is now set in subnet-up/down scripts.

 * Tinc now gives an error when unknown options are given on the command line.

 * Tinc now correctly handles a space between a short command line option and
   an optional argument.

Thanks to Etienne Dechamps for his contribution to this version of tinc.

Version 1.0.21               April 22 2013

 * Drop packets forwarded via TCP if they are too big (CVE-2013-1428).


src/tincd.c
/*
    tincd.c -- the main file for tincd
    Copyright (C) 1998-2005 Ivo Timmermans
                  2000-2022 Guus Sliepen <guus@tinc-vpn.org>
                  2008      Max Rijevski <maksuf@gmail.com>
                  2009      Michael Tokarev <mjt@tls.msk.ru>
                  2010      Julien Muchembled <jm@jmuchemb.eu>
                  2010      Timothy Redaelli <timothy@redaelli.eu>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include "system.h"

/* Darwin (MacOS/X) needs the following definition... */
#ifndef _P1003_1B_VISIBLE
#define _P1003_1B_VISIBLE
#endif

#ifdef HAVE_LZO
#include LZO1X_H
#endif

#ifdef HAVE_LZ4
#include <lz4.h>
#endif

#ifndef HAVE_WINDOWS
#include <pwd.h>
#include <grp.h>
#include <time.h>
#endif

#include "conf.h"
#include "crypto.h"
#include "event.h"
#include "logger.h"
#include "names.h"
#include "net.h"
#include "process.h"
#include "protocol.h"
#include "utils.h"
#include "xalloc.h"
#include "version.h"
#include "random.h"
#include "sandbox.h"
#include "watchdog.h"
#include "address_cache.h"
#include "fs.h"

/* If nonzero, display usage information and exit. */
static bool show_help = false;

/* If nonzero, print the version on standard output and exit.  */
static bool show_version = false;

#ifdef HAVE_MLOCKALL
/* If nonzero, disable swapping for this process. */
static bool do_mlock = false;
#endif

#ifndef HAVE_WINDOWS
/* If nonzero, chroot to netdir after startup. */
static bool do_chroot = false;

/* If !NULL, do setuid to given user after startup */
static const char *switchuser = NULL;
#endif

char **g_argv;                  /* a copy of the cmdline arguments */

static int status = 1;

typedef enum option_t {
	OPT_BAD_OPTION  = '?',
	OPT_LONG_OPTION =  0,

	// Short options
	OPT_CONFIG_FILE = 'c',
	OPT_NETNAME     = 'n',
	OPT_NO_DETACH   = 'D',
	OPT_DEBUG       = 'd',
	OPT_MLOCK       = 'L',
	OPT_CHROOT      = 'R',
	OPT_CHANGE_USER = 'U',
	OPT_SYSLOG      = 's',
	OPT_OPTION      = 'o',

	// Long options
	OPT_HELP        = 255,
	OPT_VERSION,
	OPT_NO_SECURITY,
	OPT_LOGFILE,
	OPT_PIDFILE,
} option_t;

static struct option const long_options[] = {
	{"config",          required_argument, NULL, OPT_CONFIG_FILE},
	{"net",             required_argument, NULL, OPT_NETNAME},
	{"no-detach",       no_argument,       NULL, OPT_NO_DETACH},
	{"debug",           optional_argument, NULL, OPT_DEBUG},
	{"mlock",           no_argument,       NULL, OPT_MLOCK},
	{"chroot",          no_argument,       NULL, OPT_CHROOT},
	{"user",            required_argument, NULL, OPT_CHANGE_USER},
	{"syslog",          no_argument,       NULL, OPT_SYSLOG},
	{"option",          required_argument, NULL, OPT_OPTION},
	{"help",            no_argument,       NULL, OPT_HELP},
	{"version",         no_argument,       NULL, OPT_VERSION},
	{"bypass-security", no_argument,       NULL, OPT_NO_SECURITY},
	{"logfile",         optional_argument, NULL, OPT_LOGFILE},
	{"pidfile",         required_argument, NULL, OPT_PIDFILE},
	{NULL,              0,                 NULL, 0},
};

#ifdef HAVE_WINDOWS
static struct WSAData wsa_state;
int main2(int argc, char **argv);
#endif

static void usage(bool status) {
	if(status)
		fprintf(stderr, "Try `%s --help\' for more information.\n",
		        program_name);
	else {
		fprintf(stdout,
		        "Usage: %s [option]...\n"
		        "\n"
		        "  -c, --config=DIR              Read configuration options from DIR.\n"
		        "  -D, --no-detach               Don't fork and detach.\n"
		        "  -d, --debug[=LEVEL]           Increase debug level or set it to LEVEL.\n"
		        "  -n, --net=NETNAME             Connect to net NETNAME.\n"
#ifdef HAVE_MLOCKALL
		        "  -L, --mlock                   Lock tinc into main memory.\n"
#endif
		        "      --logfile[=FILENAME]      Write log entries to a logfile.\n"
		        "  -s  --syslog                  Use syslog instead of stderr with --no-detach.\n"
		        "      --pidfile=FILENAME        Write PID and control socket cookie to FILENAME.\n"
		        "      --bypass-security         Disables meta protocol security, for debugging.\n"
		        "  -o, --option[HOST.]KEY=VALUE  Set global/host configuration value.\n"
#ifndef HAVE_WINDOWS
		        "  -R, --chroot                  chroot to NET dir at startup.\n"
		        "  -U, --user=USER               setuid to given USER at startup.\n"
#endif
		        "      --help                    Display this help and exit.\n"
		        "      --version                 Output version information and exit.\n"
		        "\n"
		        "Report bugs to tinc@tinc-vpn.org.\n",
		        program_name);
	}
}

// Try to resolve path to absolute, return a copy of the argument if this fails.
static char *get_path_arg(char *arg) {
	char *result = absolute_path(arg);

	if(!result) {
		result = xstrdup(arg);
	}

	return result;
}

static bool parse_options(int argc, char **argv) {
	config_t *cfg;
	int r;
	int option_index = 0;
	int lineno = 0;

	while((r = getopt_long(argc, argv, "c:DLd::n:so:RU:", long_options, &option_index)) != EOF) {
		switch((option_t) r) {
		case OPT_LONG_OPTION:
			break;

		case OPT_BAD_OPTION:
			usage(true);
			goto exit_fail;

		case OPT_CONFIG_FILE:
			assert(optarg);
			free(confbase);
			confbase = get_path_arg(optarg);
			break;

		case OPT_NO_DETACH:
			do_detach = false;
			break;

		case OPT_MLOCK: /* lock tincd into RAM */
#ifndef HAVE_MLOCKALL
			logger(DEBUG_ALWAYS, LOG_ERR, "The %s option is not supported on this platform.", argv[optind - 1]);
			goto exit_fail;
#else
			do_mlock = true;
			break;
#endif

		case OPT_DEBUG: /* increase debug level */
			if(!optarg && optind < argc && *argv[optind] != '-') {
				optarg = argv[optind++];
			}

			if(optarg) {
				debug_level = atoi(optarg);
			} else {
				debug_level++;
			}

			break;

		case OPT_NETNAME:
			assert(optarg);
			free(netname);
			netname = xstrdup(optarg);
			break;

		case OPT_SYSLOG:
			use_logfile = false;
			use_syslog = true;
			break;

		case OPT_OPTION:
			cfg = parse_config_line(optarg, NULL, ++lineno);

			if(!cfg) {
				goto exit_fail;
			}

			list_insert_tail(&cmdline_conf, cfg);
			break;

#ifdef HAVE_WINDOWS

		case OPT_CHANGE_USER:
		case OPT_CHROOT:
			logger(DEBUG_ALWAYS, LOG_ERR, "The %s option is not supported on this platform.", argv[optind - 1]);
			goto exit_fail;
#else

		case OPT_CHROOT:
			do_chroot = true;
			break;

		case OPT_CHANGE_USER:
			switchuser = optarg;
			break;
#endif

		case OPT_HELP:
			show_help = true;
			break;

		case OPT_VERSION:
			show_version = true;
			break;

		case OPT_NO_SECURITY:
			bypass_security = true;
			break;

		case OPT_LOGFILE:
			use_syslog = false;
			use_logfile = true;

			if(!optarg && optind < argc && *argv[optind] != '-') {
				optarg = argv[optind++];
			}

			if(optarg) {
				free(logfilename);
				logfilename = get_path_arg(optarg);
			}

			break;

		case OPT_PIDFILE:
			assert(optarg);
			free(pidfilename);
			pidfilename = get_path_arg(optarg);
			break;

		default:
			break;
		}
	}

	if(optind < argc) {
		fprintf(stderr, "%s: unrecognized argument '%s'\n", argv[0], argv[optind]);
		usage(true);
		goto exit_fail;
	}

	if(!netname) {
		char *env_value = getenv("NETNAME");

		if(env_value) {
			netname = xstrdup(env_value);
		}
	}

	/* netname "." is special: a "top-level name" */

	if(netname && (!*netname || !strcmp(netname, "."))) {
		free(netname);
		netname = NULL;
	}

	if(netname && !check_netname(netname, false)) {
		fprintf(stderr, "Invalid character in netname!\n");
		goto exit_fail;
	}

	if(netname && !check_netname(netname, true)) {
		fprintf(stderr, "Warning: unsafe character in netname!\n");
	}

	return true;

exit_fail:
	free_names();
	list_empty_list(&cmdline_conf);
	return false;
}

static bool read_sandbox_level(void) {
	sandbox_level_t level;
	char *value = NULL;

	if(get_config_string(lookup_config(&config_tree, "Sandbox"), &value)) {
		if(!strcasecmp("off", value)) {
			level = SANDBOX_NONE;
		} else if(!strcasecmp("normal", value)) {
			level = SANDBOX_NORMAL;
		} else if(!strcasecmp("high", value)) {
			level = SANDBOX_HIGH;
		} else {
			logger(DEBUG_ALWAYS, LOG_ERR, "Bad sandbox value %s!", value);
			free(value);
			return false;
		}

		free(value);
	} else {
#ifdef HAVE_SANDBOX
		level = SANDBOX_NORMAL;
#else
		level = SANDBOX_NONE;
#endif
	}

#ifndef HAVE_SANDBOX

	if(level > SANDBOX_NONE) {
		logger(DEBUG_ALWAYS, LOG_ERR, "Sandbox is used but is not supported on this platform");
		return false;
	}

#endif
	sandbox_set_level(level);
	return true;
}

static bool drop_privs(void) {
#ifndef HAVE_WINDOWS
	uid_t uid = 0;

	if(switchuser) {
		struct passwd *pw = getpwnam(switchuser);

		if(!pw) {
			logger(DEBUG_ALWAYS, LOG_ERR, "unknown user `%s'", switchuser);
			return false;
		}

		uid = pw->pw_uid;

		// The second parameter to initgroups on macOS requires int,
		// but __gid_t is unsigned int. There's not much we can do here.
		if(initgroups(switchuser, pw->pw_gid) != 0 || // NOLINT(bugprone-narrowing-conversions)
		                setgid(pw->pw_gid) != 0) {
			logger(DEBUG_ALWAYS, LOG_ERR, "System call `%s' failed: %s",
			       "initgroups", strerror(errno));
			return false;
		}

#ifndef __ANDROID__
// Not supported in android NDK
		endgrent();
		endpwent();
#endif
	}

	if(do_chroot) {
		tzset();        /* for proper timestamps in logs */

		if(chroot(confbase) != 0 || chdir("/") != 0) {
			logger(DEBUG_ALWAYS, LOG_ERR, "System call `%s' failed: %s",
			       "chroot", strerror(errno));
			return false;
		}

		free(confbase);
		confbase = xstrdup("");
	}

	if(switchuser)
		if(setuid(uid) != 0) {
			logger(DEBUG_ALWAYS, LOG_ERR, "System call `%s' failed: %s",
			       "setuid", strerror(errno));
			return false;
		}

#endif // HAVE_WINDOWS

	makedirs(DIR_CACHE | DIR_HOSTS | DIR_INVITATIONS);

	return sandbox_enter();
}

#ifdef HAVE_WINDOWS
# define setpriority(level) !SetPriorityClass(GetCurrentProcess(), (level))

static void stop_handler(void *data, int flags) {
	(void)data;
	(void)flags;

	event_exit();
}

static BOOL WINAPI console_ctrl_handler(DWORD type) {
	(void)type;

	logger(DEBUG_ALWAYS, LOG_NOTICE, "Got console shutdown request");

	if(WSASetEvent(stop_io.event) == FALSE) {
		abort();
	}

	return TRUE;
}
#else
# define NORMAL_PRIORITY_CLASS 0
# define BELOW_NORMAL_PRIORITY_CLASS 10
# define HIGH_PRIORITY_CLASS -10
# define setpriority(level) (setpriority(PRIO_PROCESS, 0, (level)))
#endif

static void cleanup(void) {
	splay_empty_tree(&config_tree);
	list_empty_list(&cmdline_conf);
	free_names();
}

int main(int argc, char **argv) {
	program_name = argv[0];

	if(!parse_options(argc, argv)) {
		return 1;
	}

	if(show_version) {
		fprintf(stdout,
		        "%s version %s (built %s %s, protocol %d.%d)\n"
		        "Features:"
#ifdef HAVE_OPENSSL
		        " openssl"
#endif
#ifdef HAVE_LIBGCRYPT
		        " libgcrypt"
#endif
#ifdef HAVE_LZO
		        " comp_lzo"
#endif
#ifdef HAVE_ZLIB
		        " comp_zlib"
#endif
#ifdef HAVE_LZ4
		        " comp_lz4"
#endif
#ifndef DISABLE_LEGACY
		        " legacy_protocol"
#endif
#ifdef ENABLE_JUMBOGRAMS
		        " jumbograms"
#endif
#ifdef ENABLE_TUNEMU
		        " tunemu"
#endif
#ifdef ENABLE_VMNET
		        " vmnet"
#endif
#ifdef HAVE_MINIUPNPC
		        " miniupnpc"
#endif
#ifdef HAVE_SANDBOX
		        " sandbox"
#endif
#ifdef ENABLE_UML
		        " uml"
#endif
#ifdef ENABLE_VDE
		        " vde"
#endif
#ifdef HAVE_WATCHDOG
		        " watchdog"
#endif
		        "\n\n"
		        "Copyright (C) 1998-2021 Ivo Timmermans, Guus Sliepen and others.\n"
		        "See the AUTHORS file for a complete list.\n"
		        "\n"
		        "tinc comes with ABSOLUTELY NO WARRANTY.  This is free software,\n"
		        "and you are welcome to redistribute it under certain conditions;\n"
		        "see the file COPYING for details.\n",
		        PACKAGE, BUILD_VERSION, BUILD_DATE, BUILD_TIME, PROT_MAJOR, PROT_MINOR);
		return 0;
	}

	if(show_help) {
		usage(false);
		return 0;
	}

	make_names(true);
	atexit(cleanup);

	if(chdir(confbase) == -1) {
		logger(DEBUG_ALWAYS, LOG_ERR, "Could not change to configuration directory: %s", strerror(errno));
		return 1;
	}

#ifdef HAVE_WINDOWS

	if(WSAStartup(MAKEWORD(2, 2), &wsa_state)) {
		logger(DEBUG_ALWAYS, LOG_ERR, "System call `%s' failed: %s", "WSAStartup", winerror(GetLastError()));
		return 1;
	}

#else
	// Check if we got an umbilical fd from the process that started us
	char *umbstr = getenv("TINC_UMBILICAL");

	if(umbstr) {
		int colorize = 0;
		sscanf(umbstr, "%d %d", &umbilical, &colorize);
		umbilical_colorize = colorize;

		if(fcntl(umbilical, F_GETFL) < 0) {
			umbilical = 0;
		}

#ifdef FD_CLOEXEC

		if(umbilical) {
			fcntl(umbilical, F_SETFD, FD_CLOEXEC);
		}

#endif
	}

#endif

	openlogger("tinc", use_logfile ? LOGMODE_FILE : LOGMODE_STDERR);

	g_argv = argv;

	const char *listen_pid = getenv("LISTEN_PID");

	if(listen_pid && atoi(listen_pid) == getpid()) {
		do_detach = false;
	}

#ifdef HAVE_UNSETENV
	unsetenv("LISTEN_PID");
#endif

	gettimeofday(&now, NULL);
	random_init();
	crypto_init();
	prng_init();

	resolve_init();

	if(!read_server_config(&config_tree)) {
		return 1;
	}

	if(!read_sandbox_level()) {
		return 1;
	}

	if(debug_level == DEBUG_NOTHING) {
		int level = 0;

		if(get_config_int(lookup_config(&config_tree, "LogLevel"), &level)) {
			debug_level = level;
		}
	}

#ifdef HAVE_LZO

	if(lzo_init() != LZO_E_OK) {
		logger(DEBUG_ALWAYS, LOG_ERR, "Error initializing LZO compressor!");
		return 1;
	}

#endif

#ifdef HAVE_WINDOWS
	io_add_event(&stop_io, stop_handler, NULL, WSACreateEvent());

	if(stop_io.event == FALSE) {
		abort();
	}

	int result;

	if(!do_detach || !init_service()) {
		SetConsoleCtrlHandler(console_ctrl_handler, TRUE);
		result = main2(argc, argv);
	} else {
		result = 1;
	}

	if(WSACloseEvent(stop_io.event) == FALSE) {
		abort();
	}

	io_del(&stop_io);
	return result;
}

int main2(int argc, char **argv) {
	(void)argc;
	(void)argv;
#endif
	char *priority = NULL;

	if(!detach()) {
		return 1;
	}

#ifdef HAVE_MLOCKALL

	/* Lock all pages into memory if requested.
	 * This has to be done after daemon()/fork() so it works for child.
	 * No need to do that in parent as it's very short-lived. */
	if(do_mlock && mlockall(MCL_CURRENT | MCL_FUTURE) != 0) {
		logger(DEBUG_ALWAYS, LOG_ERR, "System call `%s' failed: %s", "mlockall",
		       strerror(errno));
		return 1;
	}

#endif

	/* Setup sockets and open device. */

	if(!setup_network()) {
		goto end;
	}

	/* Change process priority */

	if(get_config_string(lookup_config(&config_tree, "ProcessPriority"), &priority)) {
		if(!strcasecmp(priority, "Normal")) {
			if(setpriority(NORMAL_PRIORITY_CLASS) != 0) {
				logger(DEBUG_ALWAYS, LOG_ERR, "System call `%s' failed: %s", "setpriority", strerror(errno));
				goto end;
			}
		} else if(!strcasecmp(priority, "Low")) {
			if(setpriority(BELOW_NORMAL_PRIORITY_CLASS) != 0) {
				logger(DEBUG_ALWAYS, LOG_ERR, "System call `%s' failed: %s", "setpriority", strerror(errno));
				goto end;
			}
		} else if(!strcasecmp(priority, "High")) {
			if(setpriority(HIGH_PRIORITY_CLASS) != 0) {
				logger(DEBUG_ALWAYS, LOG_ERR, "System call `%s' failed: %s", "setpriority", strerror(errno));
				goto end;
			}
		} else {
			logger(DEBUG_ALWAYS, LOG_ERR, "Invalid priority `%s`!", priority);
			goto end;
		}
	}

	/* drop privileges */
	if(!drop_privs()) {
		goto end;
	}

	/* Start main loop. It only exits when tinc is killed. */

	logger(DEBUG_ALWAYS, LOG_NOTICE, "Ready");

	if(umbilical) { // snip!
		if(write(umbilical, "", 1) != 1) {
			// Pipe full or broken, nothing we can do about it.
		}

		close(umbilical);
		umbilical = 0;
	}

	try_outgoing_connections();

#ifdef HAVE_WATCHDOG
	watchdog_start();
#endif

	status = main_loop();

#ifdef HAVE_WATCHDOG
	watchdog_stop();
#endif

	/* Shutdown properly. */

end:
	close_network_connections();

	logger(DEBUG_ALWAYS, LOG_NOTICE, "Terminating");

	free(priority);

	random_exit();

	return status;
}
